File: cplusplus/design_pattern/ObserverProj/SubjectNew.cpp
#include "stdafx.h"
#include "SubjectNew.h"


SubjectNew::SubjectNew()
{
}


SubjectNew::~SubjectNew()
{
}

void SubjectNew::notifyAll(std::string msg)
{
    std::for_each(collect.begin(), collect.end(), [msg](auto it) { it->update(msg); });
}

void SubjectNew::addObserver(std::shared_ptr<Observer> ob)
{
    if (!(std::find(collect.begin(), collect.end(), ob) != collect.end())) {
        collect.push_back(ob);
    }
}


void SubjectNew::deleteObserver(std::shared_ptr<Observer> ob)
{
    auto it = collect.begin();
    if ((it = std::find(collect.begin(), collect.end(), ob)) != collect.end()) {
        collect.erase(it);
    }
}

File: Java/pattern/ObserverProj/src/com/pattern/SubjectNew.java
package com.pattern;

import java.util.ArrayList;

public class SubjectNew {
	private ArrayList<Observer> collect = new ArrayList<Observer>();
	
	public void addObserver(Observer ob) {
		if (!collect.contains(ob)) {
			collect.add(ob);
		}
	}
	
	public void deleteObServer(Observer ob) {
		if (collect.contains(ob))
			collect.remove(ob);
	}
	
	public void notifyAll(String msg) {
		int arrayLen = collect.size();
		for (int i = 0; i < arrayLen; ++i) {
			collect.get(i).update(msg);
		}
	}
	
}
